### ISSUE TITLE
Plover not working with Dvorak layout on Debian Linux.

### ISSUE THREAD

**Issue by tahti** [openstenoproject/plover#788]
Classification: Bug

Reproducibility: Always

## Summary

When using Plover with Dvorak layout, words are printed as if Plover thought there is Dvorak layout but actually Qwerty was used. For example "g;" is printed instead of "is". Any word is changed in similar way.

## Steps to Reproduce

1. Run plover
2. Press `a` key

## Expected Results

3. Word `is` should be printed.

## Actual Results

3. Word `g;` is printed.

## Version

Plover version 4.0.0.dev1+20.g60a373f 

Run via: AppImage.

## Notes
The bug seems to be a regression bug as running Plover 3.0.0 on exactly the same configuration works correctly. I also noticed that `fbxkb` changes flag as if keyboard layout was changed to qwerty (even though I have no qwerty in my configuration files).

I am not sure if it is relevant but xdotool seems to fail in similar way:
   xdotool key "i"
produces `g`

## Configuration

OS: Linux, Debian Testing with i3 as window layout manager.
Content of `/etc/default/keyboard`:

    XKBMODEL="pc105"
    XKBLAYOUT="pl"
    XKBVARIANT="dvp"
    XKBOPTIONS="caps:swapescape,group:sclk_toggle"

uname -a output: `Linux 4.11.0-1-amd64 #1 SMP Debian 4.11.6-1 (2017-06-19) x86_64 GNU/Linux`

**Comment by Quetzal2** [comment]
I confirm the bug. It happens with plover-4.0.0.dev1.20.g60a373f-x86_64.AppImage and 3.1.1 , under Manjaro Linux. I did not try further versions.

umane -a output : `4.9.35-1-MANJARO #1 SMP PREEMPT Fri Jun 30 06:42:12 UTC 2017 x86_64 GNU/Linux`

**Comment by nimble0** [comment]
This might be the same issue as https://github.com/openstenoproject/plover/issues/650

**Comment by Quetzal2** [comment]
Ok, so I did some troubleshooting, and figured out that **Plover always uses the _first keyboard layout_ set with `setxkbmap`**.

Lets take for example this setup:

`setxkbmap us,us ,colemak -option -option grp:sclk_toggle`        _Qwerty first and colemak second, to toggle, use scroll lock key._

To check the layouts state, I use `setxkbmap -print` :

**Comment by benoit-pierre** [comment]
Yes, as noted in #650, xkb groups are not supported.

**Comment by benoit-pierre** [comment]
Duplicate of #650.

**Comment by Quetzal2** [comment]
Ok, so I did some troubleshooting, and figured out that **① Plover always uses the _first keyboard layout_ set with `setxkbmap` ② at Plovers startup**.

Lets take for example this setup:

`setxkbmap us,us ,colemak -option -option grp:sclk_toggle`        _(Qwerty first and colemak second, to toggle, use scroll lock key.)_

To check the layouts state, I use `setxkbmap -print` :
`xkb_keymap {
        xkb_keycodes  { include "evdev+aliases(qwerty)" };
        xkb_types     { include "complete"      };
        xkb_compat    { include "complete"      };
        xkb_symbols   { include "pc+us+us(colemak):2+inet(evdev)+group(sclk_toggle)"    };
        xkb_geometry  { include "pc(pc101)"     };
};`

Once the command's run, open plover.
With this setup, I get (typing the demo sentence http://stenoknight.com/kws.html => SO S TH WOG HF):
`so is this working?`

By changing the layout to colemak (Scroll lock key), I get (plover still open):
`ry ur ghur wypeukd?`

So that's one thing.

But when changing the layout to qwerty only (`setxkbmap us -option`)
and then opening plover, as expected, we get:
`so is this working?`

Let's change the layout to colemak only (keeping plover open):
`setxkbmap us colemak -option`

We then get:
`ry ur ghur wypeukd?`

That's it!
And don't forget to `setxkbmap -print` to see what plover reads from setxkbmap.

What do you think of both points? (by now, @benoit-pierre just answered the ①)

**Comment by benoit-pierre** [comment]
I believe both points are already mentioned in #650.

**Comment by Quetzal2** [comment]
yes. I was about to update the comment to mention this :·)

And the issue #650 is DE independent. So, everybody using plover under Linux is concerned.

**Comment by benoit-pierre** [comment]
Well, not everybody use XKB groups...

**Comment by Quetzal2** [comment]
Yes, I meant as soon as you change the layout (while plover runs, with or without groups), you're concerned.

**Comment by tahti** [comment]
I do not change layout, I do not use groups, I even deleted `group:sclk_toggle` from my configuration and I am still affected by the bug. I played a bit and I think that somehow the XTEST keyboard is set to qwerty. But `setxkbmap -device 5 -query` gives me (device 5 is XTEST keyboard):

    rules:      evdev      
    model:      pc105
    layout:     pl
    variant:    dvp
    options:    caps:swapescape

When I run  `setxkbmap -device 5` Plover starts working properly, even though the result of `setxkbmap -device 5 -query` does not change. I have an impression that it may be bug in XTEST keyboard. Is Plover 3.0.0 working because does not uses XTEST keyboard?

**Comment by benoit-pierre** [comment]
Yes, the code was changed to use the XTEST keyboard. What exact command do you use to configure your keyboard layout?

**Comment by tahti** [comment]
I use the configuration file `/etc/default/keyboard ` which I believe is more or less equivalent to using `setxkbmap ` but additionally it affects the `tty` terminals. So I do not use any command, just that configuration file.

**Comment by tahti** [comment]
The think that might have messed up XTEST keyboard is that before I had three layouts in that file (including qwerty), but recently I changed to using only one Dvorak layout.

**Comment by benoit-pierre** [comment]
I have this [ugly hack](https://github.com/benoit-pierre/config-x11/blob/master/bin/setxinput.sh#L171) in my script to configure X11 input, I wonder if it's the same issue (and if it's still relevant, needs more testing).

**Comment by kwutchak** [comment]
I experienced the same issue.

Ubuntu 24.04.2, plover-4.0.0rc2+6.g53c416f-x86_64.AppImage, dotool

I managed to resolve it by configuring dotool to use the correct keyboard.   I found these option in dotool's man page.

```
setenv DOTOOL_XKB_LAYOUT us 
setenv DOTOOL_XKB_VARIANT dvorak

cd ${HOME}
./plover-4.0.0rc2+6.g53c416f-x86_64.AppImage

```

Those using "real" shells should probably use DOTOOL_XKB_LAYOUT=us etc. :)

I found my way here as I noted that plover itself output the correct text.   The text entered into the text box in focus did not match what plover wanted.

I'm guessing there's a way to determine the global keyboard settings from the CLI and setting these automatically.   That's my next move.
